src/storage/sql/migrations/versions/1bf4e4e7b24f_initial_schema.py

```diff
@@ -0,0 +1,103 @@
+"""Initial schema
+
+Revision ID: 1bf4e4e7b24f
+Revises:
+Create Date: 2020-07-26 08:46:28.752972
+
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+revision = '1bf4e4e7b24f'
+down_revision = None
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.create_table(
+        'snippets',
+
+        # an internal autoincrementing identifier used in foreign keys.
+        # Normally not visible publicly, except for the case when it is
+        # used for looking legacy snippets up by id
+        sa.Column('id', sa.Integer, primary_key=True),
+        # a short unique snippet identifier visible to users
+        sa.Column('slug', sa.String(32), nullable=False),
+
+        sa.Column('title', sa.Text),
+        sa.Column('syntax', sa.Text),
+        sa.Column('created_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+        sa.Column('updated_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+
+        # slugs must be unique (this will also automatically create a unique index)
+        sa.UniqueConstraint('slug', name='uq_slug'),
+
+        # will be used for pagination; slug guarantees uniqueness of the sorting key
+        sa.Index('snippets_created_at_slug', 'created_at', 'slug'),
+        sa.Index('snippets_updated_at_slug', 'updated_at', 'slug'),
+    )
+
+    op.create_table(
+        'changesets',
+
+        sa.Column('id', sa.Integer, primary_key=True),
+        sa.Column('snippet_id', sa.Integer, nullable=False),
+
+        # numeric index used to determine the ordering of changesets for a given snippet
+        sa.Column('version', sa.Integer,
+                  server_default=sa.text('0'), nullable=False),
+        sa.Column('content', sa.Text, nullable=False),
+        sa.Column('created_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+        sa.Column('updated_at', sa.DateTime(timezone=True),
+                  server_default=sa.func.now(), nullable=False),
+
+        # each changeset is supposed to have a unique version number
+        sa.UniqueConstraint('snippet_id', 'version',
+                            name='uq_version'),
+        # sanity check: do not allow empty changesets
+        sa.CheckConstraint('LENGTH(content) > 0',
+                           name='check_not_empty'),
+        # sanity check: version numbers are non-negative integers
+        sa.CheckConstraint('version >= 0',
+                           name='check_non_negative_version')
+    )
+    with op.batch_alter_table('changesets') as batch_op:
+        # there can be multiple changesets per snippet; changesets should be
+        # deleted when the parent snippet is deleted
+        batch_op.create_foreign_key('fk_snippet', 'snippets',
+                                    ['snippet_id'], ['id'],
+                                    ondelete='CASCADE')
```

Comment:
**Q:** For my education, can you please explain why `batch_alter_table` is used? Can't we just create a foreign key sing `op.` or even as a part of `op.create_table` ?

Reply:
This is the first time I've used this as well. The short answer is that this is only needed to support SQLite :) You can't ALTER tables in SQLite, so what Alembic does is that it creates a new table under the hood and copies all existing rows there. My understanding is that they restrict the interface of  `create_foreign_key()` to `batch_alter_table()` specifically so that all modifications are batched and the copy is only performed once.

Reply:
Hm, I think I'm still missing the point. Are you saying that if I run this migration with SQLlite engine, there will be no `op.create_foreign_key`? Also, even if SQLite does not support ALTER, can't we define a first time foreign key constraint in the same SQL query that creates a table?

Reply:
> Are you saying that if I run this migration with SQLlite engine, there will be no op.create_foreign_key? 

There will be. But it will fail at runtime and point you to `batch_alter_table()` docs.

> Also, even if SQLite does not support ALTER, can't we define a first time foreign key constraint in the same SQL query that creates a table?

We totally can. I just could not remember how to do it off the top of my head :) I'll update the code.